### PR TITLE
types: add typings for element(...).getAttributes()

### LINKED ISSUE #2971: Add getAttributes description to typescript types
I could not find getAttributes method in the latest version. I thought getAttributes was supported on both Android and iOS. Please let me know if it has been removed? Are there any alternatives in that case.

### PR DESCRIPTION
**Pull request by princefishthrower**
## Description

- This pull request resolves #2971 

In this pull request, I have added the `getAttributes()` TypeScript typings as requested - not sure the exact styling requirements of the `index.d.ts` file - I'd be happy to rearrange anything if needed.

- [x] add an E2E test for a scroll view
- [x] add an E2E test for a scroll view child
- [x] add an E2E test for multiple views
- [x] verify that typings are **accurate** after all we see in the E2E tests

> _For features/enhancements:_
 - [X] I have added/updated the relevant references in the [documentation](https://github.com/wix/Detox/tree/master/docs) files.

> _For API changes:_
 - [X] I have made the necessary changes in the [types index](https://github.com/wix/Detox/blob/master/detox/index.d.ts) file.

### CODE REVIEW

**Comment by d4vidi** [comment]
This is awesome. Thank you. We will review by next week.

**Comment by princefishthrower** [comment]
@d4vidi any update here?

**Comment by d4vidi** [comment]
Will do my best to merge tomorrow

**Comment by princefishthrower** [comment]
Great! Much appreciated. It would also be great to get the `hacktoberfest-accepted` too - this would be my fourth approved pull request for the challenge!

**Comment by Gnarwall19** [comment]
@princefishthrower  this is amazing! If you're still actively working on this, would you be able to add the optional point/coordinate parameters on the [`tap()`](https://github.com/wix/Detox/blob/master/docs/APIRef.ActionsOnElement.md#tappoint) method? I know it's technically out of scope for this, but it would be nice to not have to use the depreciated [`tapAtPoint()`](https://github.com/wix/Detox/blob/master/docs/APIRef.ActionsOnElement.md#tappoint) method to avoid typescript yelling at me.

**Comment by noomorph** [comment]
@Gnarwall19 it is already GAed in the most recent Detox version, you can use `tap({ x, y })`.

**Comment by princefishthrower** [comment]
@d4vidi - I'm not really sure how to rebase, so took the current `index.d.ts` from master, checked the diff, and then added back in my typings... let me know if this works or if I've just made a mess of things :)

**Comment by noomorph** [comment]
I'm going to take and finish this during the next 1-2 days.

File: detox/index.d.ts [hunk 1]
```diff
@@ -1376,6 +1394,149 @@ declare global {
...
+            contentOffset: number;
+            /**
+             * the content inset (in case the element is a scroll view) (iOS Only)
+             */
+            contentInset: number;
+            /**
+             * the adjusted content inset (in case the element is a scroll view) (iOS Only)
+             */
+            adjustedContentInset: number;
+            layer: string;
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/index.d.ts:1534]: i'll check

**Comment by d4vidi** [comment]
@noomorph do you want to bugathon-it?

**Comment by noomorph** [comment]
@d4vidi yes, I'll resume it tomorrow.  If you remember from yesterday, I had a problem with my local Android env, which dramatically delayed reviewing this PR as well.

**Comment by noomorph** [comment]
😌 finally.

**Comment by noomorph** [comment]
@princefishthrower it took long, but the pull request was very problematic: there were incorrect assumptions about types, so I had to verify thoroughly every property, every type. Thanks for starting the initiative! Next week Detox will have `getAttributes()` released.

cc @d4vidi

File: detox/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/index.d.ts:1570]: Yes, it is. Closing the conversation.

**Comment by noomorph** [comment]
@d4vidi , @jonathanmos, @asafkorem if anyone of you has time to take a look one more time on it, you are welcome.

**Review by noomorph — APPROVED** [review]

File: detox/index.d.ts [hunk 2]
```diff
@@ -1445,6 +1462,149 @@ declare global {
             };
         }
 
+        // Element Attributes Shared Among iOS and Android
+        interface ElementAttributes {
```
Comment (asafkorem) [inline detox/index.d.ts:1466]: Just to point out, it seems like there are more attributes that we can provide with this API (see https://reactnative.dev/docs/accessibility), like `accessibilityHint`, or `accessibilityState`..

File: detox/test/types/detox-global-tests.ts [hunk 3]
```diff
@@ -93,4 +93,22 @@ describe("Test", () => {
...
+          iosAttribute = attributes.elements[0];
+        } else if ('activationPoint' in attributes) {
+          iosAttribute = attributes;
+        } else {
+          androidAttribute = attributes;
+        }
+
+        if (!iosAttribute && !androidAttribute) {
+            throw new Error('error');
+        }
```
Comment (asafkorem) [inline detox/test/types/detox-global-tests.ts:112]: why do we throw here if we already assigned it to `androidAttribute` in the else block?

File: detox/test/types/detox-global-tests.ts [hunk 4]
```diff
@@ -93,4 +93,22 @@ describe("Test", () => {
...
         await expect(webview.element(by.web.cssSelector(".button")).atIndex(1)).toExist();
     });
+
+    test("getAttributes", async () => {
+        let iosAttribute: Detox.IosElementAttributes | undefined;
+        let androidAttribute: Detox.AndroidElementAttributes | undefined;
+
+        const attributes = await element(by.id("element")).getAttributes();
+        if ('elements' in attributes) {
+          iosAttribute = attributes.elements[0];
```
Comment (asafkorem) [inline detox/test/types/detox-global-tests.ts:103]: is that possible?

File: detox/test/types/detox-global-tests.ts [hunk 5]
```diff
@@ -93,4 +93,22 @@ describe("Test", () => {
         await expect(webview.element(by.web.cssSelector(".button"))).toExist();
         await expect(webview.element(by.web.cssSelector(".button")).atIndex(1)).toExist();
     });
+
+    test("getAttributes", async () => {
```
Comment (asafkorem) [inline detox/test/types/detox-global-tests.ts:97]: we should split this test into three different tests, for each attrs type.

File: detox/test/types/detox-global-tests.ts (hunk 3 above)
Comment (asafkorem) [reply detox/test/types/detox-global-tests.ts:112]: anyway I think we should split this test.

**Review by asafkorem — CHANGES_REQUESTED** [review]
Fantastic job! 🤩  🚀 
Some minor comments..

File: detox/index.d.ts [hunk 6]
```diff
@@ -1227,6 +1226,24 @@ declare global {
...
+             *    const jestExpect = require('expect');
+             *    // 'visible' attribute available on both iOS and Android
+             *    jestExpect(attributes.visible).toBe(true);
+             *    // 'activationPoint' attribute available on iOS only
+             *    jestExpect(attributes.activationPoint.x).toHaveValue(50);
+             *    // 'width' attribute available on Android only
+             *    jestExpect(attributes.width).toHaveValue(100);
+             * })
+             */
+             getAttributes(): Promise<IosElementAttributes | AndroidElementAttributes | { elements: IosElementAttributes[]; }>;
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/index.d.ts:1246]: @asafkorem do you really want to add new public user-facing types for a lousy return value type? We'll add amount of type support onto ourselves. It is never reused so far, this is why I don't want to engrave these not the best decisions into our type schema.

[the remark this answers is not in the document]
Comment (noomorph) [reply detox/index.d.ts:1246]: @asafkorem when multiple elements are matched (by a non-strict matcher, e.g. `by.type("RCTView")`).

File: detox/index.d.ts (hunk 2 above)
Comment (noomorph) [reply detox/index.d.ts:1466]: Not sure, @asafkorem.

I intercepted what the iOS side returns and I see no traces of the mentioned properties:

```json
{
  "elementSafeBounds": {
    "y": 0,
    "x": 0,
    "width": 426,
    "height": 149.99996948242188
  },
  "hittable": true,
  "frame": {
    "y": 469.3333435058594,
    "x": 1,
    "width": 426,
    "height": 149.99996948242188
  },
  "normalizedActivationPoint": {
    "x": 0.5,
    "y": 0.5
  },
  "visible": true,
  "elementBounds": {
    "y": 0,
    "x": 0,
    "width": 426,
    "height": 149.99996948242188
  },
  "enabled": true,
  "layer": "<CALayer: 0x60000281b8a0>",
  "safeAreaInsets": {
    "right": 0,
    "top": 0,
    "left": 0,
    "bottom": 0
  },
  "className": "RCTScrollContentView",
  "activationPoint": {
    "x": 213,
    "y": 74.99998474121094
  },
  "elementFrame": {
    "y": 0,
    "x": 0,
    "width": 426,
    "height": 149.99996948242188
  }
}
```

If you meant we can extract them potentially, then it is out of the scope of this PR, definitely. Here we are just adding typings and making sure that they are totally correct.

File: detox/index.d.ts [hunk 7]
```diff
@@ -1227,6 +1226,24 @@ declare global {
              * });
              */
              takeScreenshot(name: string): Promise<string>;
+
+
+            /**
+             * Returns an object, representing various attributes of the element. Note that only a handful of attributes are available on both iOS and Android. Many others are OS specific.
+             * For more information, see {@link https://wix.github.io/Detox/docs/api/actions-on-element/#getattributes}
```
[the remark this answers is not in the document]
Comment (noomorph) [reply detox/index.d.ts:1233]: Hmm... I've updated a bit the signature to make it terser.

File: detox/index.d.ts (hunk 2 above)
Comment (asafkorem) [reply detox/index.d.ts:1466]: Sure, out of this PR scope. I'm talking about a change that requires both Android and iOS implementations.

File: detox/index.d.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (asafkorem) [reply detox/index.d.ts]: mm yeah, looks better 👍🏼

File: detox/test/types/detox-global-tests.ts (hunk 4 above)
Comment (noomorph) [reply detox/test/types/detox-global-tests.ts:103]: yes

File: detox/test/types/detox-global-tests.ts (hunk 3 above)
Comment (noomorph) [reply detox/test/types/detox-global-tests.ts]: Thanks, I fixed both problems.

File: detox/index.d.ts (hunk 2 above)
Comment (noomorph) [reply detox/index.d.ts:1466]: https://github.com/wix/Detox/issues/3116

**Review by asafkorem — APPROVED** [review]
LGTM

**Pull request merged**